File: Data Structures & Algorithms/trapping-rain-water/submission-4.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        vector<int> prefix(n);
        vector<int> suffix(n);
        prefix[0]=height[0];
        suffix[n-1]=height[n-1];
        for(int i=1;i<n;i++){
            prefix[i]=max(prefix[i-1],height[i]);
            suffix[n-1-i]=max(suffix[n-1-i+1],height[n-1-i]);
        }
        int trap=0;
        for(int i=0;i<n;i++){
                trap+=min(prefix[i],suffix[i])-height[i];
        }
        return trap;

    }
};

File: Data Structures & Algorithms/max-water-container/submission-4.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        n=len(heights)
        l=0
        r=n-1
        maxAmount=0
        while l<r:
            h=min(heights[l],heights[r])
            w=r-l
            a=h*w
            maxAmount=max(maxAmount,a)
            if(heights[l]<heights[r]):l+=1
            else : r-=1
        
        return maxAmount

        